pandas/core/arrays/_mixins.py

```diff
@@ -64,22 +65,24 @@ def _validate_fill_value(self, fill_value):
 
     # ------------------------------------------------------------------------
 
+    # TODO: make this a cache_readonly; for that to work we need to remove
+    #  the _index_data kludge in libreduction
     @property
     def shape(self) -> Tuple[int, ...]:
         return self._ndarray.shape
 
     def __len__(self) -> int:
         return self.shape[0]
 
-    @property
+    @cache_readonly
```

Comment:
really? these are not worth it i think

Reply:
in profiling it made a bigger difference than i expected